runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SdkComponents.java

```diff
@@ -54,6 +62,48 @@ static SdkComponents create() {
     return new SdkComponents();
   }
 
+  public static RunnerApi.Pipeline translatePipeline(Pipeline p) {
+    final SdkComponents components = create();
+    final Collection<String> rootIds = new HashSet<>();
+    p.traverseTopologically(
+        new PipelineVisitor.Defaults() {
+          private final ListMultimap<Node, AppliedPTransform<?, ?, ?>> children =
+              ArrayListMultimap.create();
+
+          @Override
+          public void leaveCompositeTransform(Node node) {
+            if (node.isRootNode()) {
+              for (AppliedPTransform<?, ?, ?> pipelineRoot : children.get(node)) {
```

Comment:
I never had a reason to revisit this, but technically this records more roots than needed since it gets all the top-level transforms whether or not they have inputs. No need for any changes; best to keep it the same.